File: als-common/shared/src/main/scala/org/mulesoft/als/configuration/AlsConfiguration.scala
package org.mulesoft.als.configuration

import org.mulesoft.lsp.configuration.{DefaultFormattingOptions, FormatOptions, FormattingOptions}

case class AlsConfiguration(private var formattingOptions: Map[String, FormattingOptions] = Map(),
                            private var templateType: TemplateTypes.TemplateTypes = TemplateTypes.FULL,
                            private var prettyPrintSerialization: Boolean = false)
    extends AlsConfigurationReader {

  private var enableUpdateFormatOptions = true
  private var sDocumentChanges: Boolean = false

  override def supportsDocumentChanges: Boolean        = this.sDocumentChanges
  def supportsDocumentChanges(supports: Boolean): Unit = this.sDocumentChanges = supports

  def getFormatOptionForMime(mimeType: String): FormatOptions =
    formattingOptions.getOrElse(mimeType, DefaultFormattingOptions)

  def updateFormattingOptions(options: Map[String, FormattingOptions]): Unit =
    if (enableUpdateFormatOptions)
      options.foreach(pair => {
        this.formattingOptions += pair
      })

  def setUpdateFormatOptions(enableUpdateFormatOptions: Boolean): Unit =
    this.enableUpdateFormatOptions = enableUpdateFormatOptions

  def getFormatOptions: Map[String, FormattingOptions] = formattingOptions

  def updateFormatOptionsIsEnabled(): Boolean = enableUpdateFormatOptions

  override def getTemplateType: TemplateTypes.TemplateTypes = this.templateType

  def setTemplateType(templateType: TemplateTypes.TemplateTypes): Unit = this.templateType = templateType

  def getShouldPrettyPrintSerialization: Boolean = prettyPrintSerialization

  def setShouldPrettyPrintSerialization(value: Boolean): Unit = this.prettyPrintSerialization = value
}

object TemplateTypes extends Enumeration {
  type TemplateTypes = String
  val NONE   = "NONE"
  val SIMPLE = "SIMPLE"
  val FULL   = "FULL"
}

File: als-server/jvm/src/main/scala/org/mulesoft/als/server/lsp4j/extension/AlsConfiguration.java
package org.mulesoft.als.server.lsp4j.extension;

import java.util.Map;

public class AlsConfiguration {

    private String templateType;

    private Map<String, AlsFormattingOptions> formattingOptions;

    public AlsConfiguration(Map<String, AlsFormattingOptions> formattingOptions, String templateType) {
        this.formattingOptions = formattingOptions;
        this.templateType = templateType;
    }

    public AlsConfiguration() {
    }

    public Map<String, AlsFormattingOptions> getFormattingOptions() {
        return formattingOptions;
    }

    public void setFormattingOptions(Map<String, AlsFormattingOptions> formattingOptions) {
        this.formattingOptions = formattingOptions;
    }

    public String getTemplateType() {
        return templateType;
    }

    public void setTemplateType(String templateType) {
        this.templateType = templateType;
    }
}
